patina_dxe_core/src/locks/tpl_mutex.rs

```diff
@@ -0,0 +1,212 @@
+//! TplMutex: Mutex implementation that also adjusts UEFI TPL levels
+//!
+//! This module uses a Mutex to provide mutual exclusion, but also raises and
+//! lowers the UEFI TPL level when the mutex is locked and unlocked.
+//!
+//! This mutex guarantees that the critical section protected by the guard
+//! cannot be interrupted by code running at TPL equal to or lower than the
+//! lock's TPL level. At TPL_HIGH_LEVEL, interrupts are disabled, so that
+//! means that the critical section cannot be interrupted by anything.
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation.
+//!
+//! SPDX-License-Identifier: Apache-2.0
+use core::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use r_efi::efi;
+
+use crate::{
+    events::{raise_tpl, restore_tpl},
+    locks::mutex::{Mutex, MutexGuard},
+};
+
+/// Used to guard data with a locked MUTEX and TPL level.
+pub struct TplMutex<T: ?Sized> {
+    tpl_lock_level: efi::Tpl,
+    data: Mutex<T>,
+}
+/// Wrapper for guarded data, which can be accessed by Deref or DerefMut on this object.
+pub struct TplGuard<'a, T: ?Sized + 'a> {
+    release_tpl: efi::Tpl,
+    guard: Option<MutexGuard<'a, T>>,
+}
+
+unsafe impl<T: ?Sized + Send> Sync for TplMutex<T> {}
+unsafe impl<T: ?Sized + Send> Send for TplMutex<T> {}
+
+unsafe impl<T: ?Sized + Sync> Sync for TplGuard<'_, T> {}
+unsafe impl<T: ?Sized + Send> Send for TplGuard<'_, T> {}
+
+impl<T> TplMutex<T> {
+    /// Instantiates a new TplMutex with the given TPL level, data object, and name string.
+    pub const fn new(tpl_lock_level: efi::Tpl, data: T, name: &'static str) -> Self {
+        Self { tpl_lock_level, data: Mutex::new(data, name) }
+    }
+}
+
+impl<T: ?Sized> TplMutex<T> {
+    /// Lock the TplMutex and return a TplGuard object used to access the data. This will raise the system TPL level
+    /// to the level specified at TplMutex creation.
+    ///
+    /// Safety: Lock reentrance is not supported; attempt to re-lock something already locked will panic.
+    pub fn lock(&self) -> TplGuard<'_, T> {
+        self.try_lock().unwrap_or_else(|| panic!("Re-entrant locks for {:?} not permitted.", self.data.name()))
+    }
+
+    /// Attempts to lock the TplMutex, and if successful, returns a guard object that can be used to access the data.
+    pub fn try_lock(&self) -> Option<TplGuard<'_, T>> {
+        let release_tpl = raise_tpl(self.tpl_lock_level);
```

Comment:
Probably a comment is worthwhile going in here, let me make sure my understanding is correct:

For all TPLs != TPL_HIGH: we use the Mutex to briefly disable interrupts while acquiring the lock, then run with interrupts enabled.

For TPL_HIGH, we do the same thing, but are relying on raise_tpl to disable interrupts and lower_tpl to lower interrupts.

i.e. that InterruptMutex guarantees the interrupt semantics, but TplMutex does not; it relies on [raise|lower]_tpl to handle the interrupts.

Reply:
Correct. TplMutex makes the guarantee that your critical section (including acquisition of the lock) is guarded against concurrency at or below the lock level specified when you create the TplMutex. This relies on the fact that (aside from special cases that should not be using TplMutex) architecturally there is only one "interrupt" - the timer tick, and handling of it is TPL-aware. 

If you tried to acquire the TplMutex at a TPL higher than specified in the lock, (i.e. you are running at a handler in TPL_NOTIFY, and you try and acquire a lock that is TPL_CALLBACK) you would panic due to TPL inversion.  

Put another way: TplMutex (except at high) don't _need_ to disable interrupts, because the TPL architecture/Timer architecture makes the necessary concurrency guarantees.